### PR TITLE
Prevent closed sparks from joining rooms

### PR DESCRIPTION
**Pull request by lpinca**
Right now a spark can join a room even if it is already closed and this can create a memory leak as shown in #25.
This patch should fix the issue.

### CODE REVIEW

File: lib/rooms.js [hunk 1]
```diff
@@ -324,25 +317,23 @@ Rooms.prototype.write = function write(data) {
...
-
-    // We need to make sure this is a valid spark and not a destroyed one,
-    // destroyed spark are no longer part of the connection list
-    var conn = this.connections[this.id];
-
-    // If this is not a valid ctx we just ignore the write request
-    if (!conn) return this.reset();
-
+    // Ignore the broadcast request if the spark is closed
+    if (!this.primus.spark(this.id)) return this.reset();
```
Comment (lpinca) [inline lib/rooms.js]: I've tried with `this.ctx.readyState === this.primus.Spark.CLOSED` here but unfortunately it doesn't work with `primus-multiplex`.

**Comment by lpinca** [comment]
I think that an error should be thrown when used synchronously and the spark is closed:
```js
spark.join('foo bar baz'); // should throw if spark is closed
```
but it doesn't now as it's not trivial to do with the current architecture.
I think we should always and only use an async interface.

**Comment by cayasso** [comment]
@lpinca thank you for the updates, so you say to check if there is no callback passed as argument then throw the error? I think that is better than nothing, currently it would fail anyway right.

BTW. I am almost done on with 1.0.x primus-rooms-adapter I will push to 1.0 branch so you can take a look

**Comment by lpinca** [comment]
> so you say to check if there is no callback passed as argument then throw the error?

Yes but it's tricky because the operation is actually [async](https://github.com/cayasso/primus-rooms/blob/bd79aad81b13b235569582ba659c33bd95809489/lib/rooms.js#L415). If the spark is closed mid way the error would not be thrown and the user would assume that all rooms have been joined successfully.

**Comment by lpinca** [comment]
Actually all the [parallel tasks](https://github.com/cayasso/primus-rooms/blob/bd79aad81b13b235569582ba659c33bd95809489/lib/rooms.js#L409-L413) are called on the same tick so they would find the spark in same ready state.

I guess it's safe to move the check from the [`_join`](https://github.com/cayasso/primus-rooms/blob/bd79aad81b13b235569582ba659c33bd95809489/lib/rooms.js#L127-L140) method to the [`join`](https://github.com/cayasso/primus-rooms/blob/bd79aad81b13b235569582ba659c33bd95809489/lib/rooms.js#L114-L116) method and throw an error if the callback is not provided.

I'll send a pr later.

**Pull request merged**